--- a/providers/ionic/ionic.h	2026-03-14 14:10:29.909261042 -0600
+++ b/providers/ionic/ionic.h	2026-03-14 14:12:09.904972186 -0600
@@ -19,6 +19,7 @@
 
 #include "ionic-abi.h"
 
+#include "ionic_dv.h"
 #include "ionic_memory.h"
 #include "ionic_queue.h"
 #include "ionic_table.h"
@@ -40,13 +41,11 @@
 #define IONIC_QUEUE_DEPTH_MAX 0xFFFF
 #define IONIC_QUEUE_STRIDE_MAX 0x10000
 
-/** IONIC_PD_TAG - tag used for parent domain resource allocation. */
-#define IONIC_PD_TAG	((uint64_t)RDMA_DRIVER_IONIC << 32)
-#define IONIC_PD_TAG_CQ	(IONIC_PD_TAG | 1)
-#define IONIC_PD_TAG_SQ	(IONIC_PD_TAG | 2)
-#define IONIC_PD_TAG_RQ	(IONIC_PD_TAG | 3)
-
 enum {
+	IONIC_CQ_SUPPORTED_COMP_MASK =
+		IBV_CQ_INIT_ATTR_MASK_FLAGS		|
+		IBV_CQ_INIT_ATTR_MASK_PD,
+
 	IONIC_CQ_SUPPORTED_WC_FLAGS =
 	    IBV_WC_EX_WITH_BYTE_LEN       |
 	    IBV_WC_EX_WITH_IMM            |
@@ -54,7 +53,13 @@
 	    IBV_WC_EX_WITH_SRC_QP         |
 	    IBV_WC_EX_WITH_SLID           |
 	    IBV_WC_EX_WITH_SL             |
-	    IBV_WC_EX_WITH_DLID_PATH_BITS
+	    IBV_WC_EX_WITH_DLID_PATH_BITS,
+
+	IONIC_CQ_SUPPORTED_EXT_COMP_MASK =
+		IONIC_CQ_INIT_ATTR_MASK_FLAGS,
+
+	IONIC_CQ_SUPPORTED_EXT_FLAGS =
+		IONIC_CQ_INIT_ATTR_CCQE,
 };
 
 struct ionic_ctx {
@@ -91,6 +96,7 @@
 	struct ibv_pd		*root_ibpd;
 
 	uint8_t			udma_mask;
+	uint8_t			expdb_mask;
 	uint8_t			sq_cmb;
 	uint8_t			rq_cmb;
 
@@ -163,6 +169,7 @@
 	uint16_t		cmb_prod;
 	uint8_t			cmb;
 	bool			flush;
+	bool			gda;
 };
 
 struct ionic_sq {
@@ -182,6 +189,7 @@
 
 	bool			flush;
 	bool			flush_rcvd;
+	bool			gda;
 	bool			color;
 };
 
@@ -291,6 +299,11 @@
 	return BIT(ctx->udma_count) - 1;
 }
 
+struct ibv_cq_ex *ionic_create_cq_ex_common(
+	struct ibv_context *ibctx,
+	struct ibv_cq_init_attr_ex *ex,
+	struct ionic_cq_init_attr_ex *ionic_ex);
+
 static inline void ionic_dbg_xdump(struct ionic_ctx *ctx, const char *str,
 				   const void *ptr, size_t size)
 {
