--- a/providers/ionic/ionic_verbs.c	2026-03-14 14:12:41.977415175 -0600
+++ b/providers/ionic/ionic_verbs.c	2026-03-14 14:13:36.586513533 -0600
@@ -548,26 +548,57 @@
 	return vcq->cur_wc.dlid_path_bits;
 }
 
-static struct ibv_cq_ex *ionic_create_cq_ex(struct ibv_context *ibctx,
-					    struct ibv_cq_init_attr_ex *ex)
+struct ibv_cq_ex *ionic_create_cq_ex_common(struct ibv_context *ibctx,
+					    struct ibv_cq_init_attr_ex *ex,
+					    struct ionic_cq_init_attr_ex *ionic_ex)
 {
 	struct ionic_ctx *ctx = to_ionic_ctx(ibctx);
 	struct ionic_pd *pd = NULL;
 	struct ionic_vcq *vcq;
 	struct uionic_cq req = {};
 	struct uionic_cq_resp resp = {};
+	uint32_t ncqe;
 	int cq_i, rc;
 
-	if (ex->wc_flags & ~IONIC_CQ_SUPPORTED_WC_FLAGS) {
+	if (ionic_ex) {
+		if (!check_comp_mask(ionic_ex->comp_mask,
+				     IONIC_CQ_SUPPORTED_EXT_COMP_MASK)) {
+			rc = ENOTSUP;
+			goto err;
+		}
+
+		if (ionic_ex->comp_mask & IONIC_CQ_INIT_ATTR_MASK_FLAGS) {
+			if (!check_comp_mask(ionic_ex->flags,
+				     IONIC_CQ_SUPPORTED_EXT_FLAGS)) {
+				rc = ENOTSUP;
+				goto err;
+			}
+		}
+	}
+
+	if (!check_comp_mask(ex->comp_mask, IONIC_CQ_SUPPORTED_COMP_MASK)) {
 		rc = ENOTSUP;
 		goto err;
 	}
 
-	if (ex->cqe < 1 || ex->cqe + IONIC_CQ_GRACE > 0xffff) {
-		rc = EINVAL;
+	if (!check_comp_mask(ex->wc_flags, IONIC_CQ_SUPPORTED_WC_FLAGS)) {
+		rc = ENOTSUP;
 		goto err;
 	}
 
+	if (ionic_ex &&
+	    (ionic_ex->comp_mask & IONIC_CQ_INIT_ATTR_MASK_FLAGS) &&
+	    (ionic_ex->flags & IONIC_CQ_INIT_ATTR_CCQE)) {
+		ncqe = 0;
+	} else {
+		ncqe = ex->cqe;
+
+		if (!ncqe || ncqe + IONIC_CQ_GRACE > 0xffff) {
+			rc = EINVAL;
+			goto err;
+		}
+	}
+
 	vcq = calloc(1, sizeof(*vcq));
 	if (!vcq) {
 		rc = ENOMEM;
@@ -653,6 +684,12 @@
 	return NULL;
 }
 
+static struct ibv_cq_ex *ionic_create_cq_ex(struct ibv_context *ibctx,
+					    struct ibv_cq_init_attr_ex *ex)
+{
+	return ionic_create_cq_ex_common(ibctx, ex, NULL);
+}
+
 static struct ibv_cq *ionic_create_cq(struct ibv_context *ibctx, int ncqe,
 				      struct ibv_comp_channel *channel,
 				      int vec)
@@ -2654,7 +2691,7 @@
 
 		if (qp->sq.cmb_ptr) {
 			ionic_post_send_cmb(ctx, qp);
-		} else {
+		} else if (!qp->sq.gda) {
 			udma_to_device_barrier();
 			verbs_debug(&ctx->vctx, "dbell qp %u sq val %" PRIx64,
 				    qp->qpid, ionic_queue_dbell_val(&qp->sq.queue));
@@ -2791,7 +2828,7 @@
 
 		if (qp->rq.cmb_ptr) {
 			ionic_post_recv_cmb(ctx, qp);
-		} else {
+		} else if (!qp->rq.gda) {
 			udma_to_device_barrier();
 			verbs_debug(&ctx->vctx, "dbell qp %u rq val %" PRIx64,
 				    qp->qpid,
